From 8533a76a0441f95365da3f95e7c6b8abdf8d2a63 Mon Sep 17 00:00:00 2001
From: Dave Stevenson <dave.stevenson@raspberrypi.org>
Date: Thu, 19 Apr 2018 09:33:30 +0100
Subject: [PATCH 54/55] v4l2_m2m: Request selection correctly at appropriate
 times.

    On format changed the selection denoting the image crop
    may well change. Reread the selection on format changed and
    update width/height appropriately.

    When setting the selection specify the target.
---
 libavcodec/v4l2_context.c | 14 ++++++++++++++
 libavcodec/v4l2_m2m_dec.c |  7 +++++--
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/libavcodec/v4l2_context.c b/libavcodec/v4l2_context.c
index fa2deae888..6b55f95830 100644
--- a/libavcodec/v4l2_context.c
+++ b/libavcodec/v4l2_context.c
@@ -165,6 +165,7 @@ static int v4l2_handle_event(V4L2Context *ctx)
     struct v4l2_format out_fmt = s->output.format;
     struct v4l2_event evt = { 0 };
     int full_reinit, reinit, ret;
+    struct v4l2_selection selection = {0};
 
     ret = ioctl(s->fd, VIDIOC_DQEVENT, &evt);
     if (ret < 0) {
@@ -204,6 +205,14 @@ static int v4l2_handle_event(V4L2Context *ctx)
         s->capture.height = v4l2_get_height(&cap_fmt);
         s->capture.width = v4l2_get_width(&cap_fmt);
         s->capture.sample_aspect_ratio = v4l2_get_sar(&s->capture);
+
+        selection.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+        selection.target = V4L2_SEL_TGT_COMPOSE;
+        ret = ioctl(s->fd, VIDIOC_G_SELECTION, &selection);
+        if (ret) {
+            av_log(logger(ctx), AV_LOG_DEBUG, "crop output %dx%d\n", selection.r.width, selection.r.height);
+            /* update the size of the resulting frame */
+        }
     }
 
     if (full_reinit || reinit)
@@ -224,6 +233,11 @@ static int v4l2_handle_event(V4L2Context *ctx)
         if (ret < 0)
             av_log(logger(ctx), AV_LOG_WARNING, "update avcodec height and width\n");
 
+        if (selection.r.width && selection.r.height) {
+            s->avctx->width  = selection.r.width;
+            s->avctx->height = selection.r.height;
+        }
+
         ret = ff_v4l2_m2m_codec_reinit(s);
         if (ret) {
             av_log(logger(ctx), AV_LOG_ERROR, "v4l2_m2m_codec_reinit\n");
diff --git a/libavcodec/v4l2_m2m_dec.c b/libavcodec/v4l2_m2m_dec.c
index 820cdf241f..66717bf2f4 100644
--- a/libavcodec/v4l2_m2m_dec.c
+++ b/libavcodec/v4l2_m2m_dec.c
@@ -75,8 +75,11 @@ static int v4l2_try_start(AVCodecContext *avctx)
 
     /* 3. set the crop parameters */
     selection.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
-    selection.r.height = avctx->coded_height;
-    selection.r.width = avctx->coded_width;
+    selection.r.height = avctx->height;
+    selection.r.width = avctx->width;
+    selection.r.left = 0;
+    selection.r.top = 0;
+    selection.target = V4L2_SEL_TGT_COMPOSE;
     ret = ioctl(s->fd, VIDIOC_S_SELECTION, &selection);
     if (!ret) {
         ret = ioctl(s->fd, VIDIOC_G_SELECTION, &selection);
-- 
2.29.2

